File: dotnet/src/Envdoctor/Parsers/YamlFacade.cs
using System.Globalization;
using System.Text.RegularExpressions;
using YamlDotNet.RepresentationModel;

namespace Envdoctor.Parsers;

/// Helpers over YamlDotNet that reproduce the scalar-resolution semantics of
/// the reference's `yaml` npm package (YAML 1.2 core schema): plain scalars
/// become null/bool/int/float/string; quoted scalars are always strings.
public static class YamlFacade
{
    private static readonly Regex IntRe = new(@"^[-+]?[0-9]+$", RegexOptions.Compiled);
    private static readonly Regex HexIntRe = new(@"^0x[0-9a-fA-F]+$", RegexOptions.Compiled);
    private static readonly Regex OctIntRe = new(@"^0o[0-7]+$", RegexOptions.Compiled);
    private static readonly Regex FloatRe = new(
        @"^[-+]?(\.[0-9]+|[0-9]+(\.[0-9]*)?)([eE][-+]?[0-9]+)?$", RegexOptions.Compiled);

    /// Parse all documents; returns an empty list when the content is not
    /// parseable YAML (the callers still scan raw text for interpolations).
    public static List<object?> LoadAll(string content)
    {
        try
        {
            var stream = new YamlStream();
            stream.Load(new StringReader(content));
            return stream.Documents.Select(d => ToPlain(d.RootNode)).ToList();
        }
        catch (YamlDotNet.Core.YamlException)
        {
            return new List<object?>();
        }
    }

    /// Parse the first document; null when unparseable.
    public static object? LoadFirst(string content)
    {
        var docs = LoadAll(content);
        return docs.Count > 0 ? docs[0] : null;
    }

    public static object? ToPlain(YamlNode node) => node switch
    {
        YamlScalarNode scalar => ResolveScalar(scalar),
        YamlSequenceNode seq => seq.Children.Select(ToPlain).ToList(),
        YamlMappingNode map => map.Children.ToDictionary(
            kv => JsString(ToPlain(kv.Key)),
            kv => ToPlain(kv.Value)),
        _ => null,
    };

    private static object? ResolveScalar(YamlScalarNode scalar)
    {
        if (scalar.Style != YamlDotNet.Core.ScalarStyle.Plain)
            return scalar.Value ?? "";
        var s = scalar.Value ?? "";
        if (s.Length == 0 || s is "~" or "null" or "Null" or "NULL")
            return null;
        if (s is "true" or "True" or "TRUE")
            return true;
        if (s is "false" or "False" or "FALSE")
            return false;
        if (IntRe.IsMatch(s) && long.TryParse(s, NumberStyles.Integer, CultureInfo.InvariantCulture, out var l))
            return l;
        if (HexIntRe.IsMatch(s) && long.TryParse(s[2..], NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var hex))
            return hex;
        if (OctIntRe.IsMatch(s))
        {
            try { return Convert.ToInt64(s[2..], 8); }
            catch (OverflowException) { }
        }
        if ((s.Contains('.') || s.Contains('e') || s.Contains('E')) &&
            FloatRe.IsMatch(s) &&
            double.TryParse(s, NumberStyles.Float, CultureInfo.InvariantCulture, out var d))
            return d;
        return s;
    }

    /// JS `String(value)` for the scalar shapes YAML resolution produces.
    public static string JsString(object? value) => value switch
    {
        null => "null",
        string s => s,
        bool b => b ? "true" : "false",
        long l => l.ToString(CultureInfo.InvariantCulture),
        double d => JsNumber(d),
        List<object?> => string.Join(",", ((List<object?>)value).Select(JsString)),
        Dictionary<string, object?> => "[object Object]",
        _ => value.ToString() ?? "",
    };

    /// JS number-to-string: shortest round-trip, lowercase `e` exponent without
    /// leading zeros, matching `Number.prototype.toString`.
    private static string JsNumber(double d)
    {
        var s = d.ToString(CultureInfo.InvariantCulture);
        var e = s.IndexOf('E');
        if (e < 0)
            return s;
        var mantissa = s[..e];
        var exp = s[(e + 1)..];
        var neg = exp.StartsWith('-');
        var digits = exp.TrimStart('-', '+').TrimStart('0');
        if (digits.Length == 0)
            digits = "0";
        return $"{mantissa}e{(neg ? "-" : "+")}{digits}";
    }
}

File: dart/lib/src/parsers/yaml_facade.dart
/// Helpers over package:yaml that reproduce the scalar-resolution semantics
/// of the reference's `yaml` npm package (YAML 1.2 core schema): plain
/// scalars become null/bool/int/float/string; quoted scalars are always
/// strings.
library;

import 'package:yaml/yaml.dart';

import '../utils/json.dart';

final RegExp _intRe = RegExp(r'^[-+]?[0-9]+$');
final RegExp _hexIntRe = RegExp(r'^0x[0-9a-fA-F]+$');
final RegExp _octIntRe = RegExp(r'^0o[0-7]+$');
final RegExp _floatRe = RegExp(r'^[-+]?(\.[0-9]+|[0-9]+(\.[0-9]*)?)([eE][-+]?[0-9]+)?$');

/// Parse all documents; returns an empty list when the content is not
/// parseable YAML (callers still scan raw text for interpolations).
List<Object?> yamlLoadAll(String content) {
  try {
    return loadYamlDocuments(content).map((d) => yamlToPlain(d.contents)).toList();
  } catch (_) {
    return <Object?>[];
  }
}

/// Parse the first document; null when unparseable.
Object? yamlLoadFirst(String content) {
  final docs = yamlLoadAll(content);
  return docs.isNotEmpty ? docs.first : null;
}

Object? yamlToPlain(YamlNode node) => switch (node) {
      YamlScalar scalar => resolveScalar(scalar),
      YamlList list => list.nodes.map(yamlToPlain).toList(),
      YamlMap map => {
          for (final entry in map.nodes.entries)
            jsString(yamlToPlain(entry.key)): yamlToPlain(entry.value),
        },
      _ => null,
    };

Object? resolveScalar(YamlScalar scalar) {
  if (scalar.style != ScalarStyle.PLAIN) return scalar.value?.toString() ?? '';
  final s = scalar.span.text;
  if (s.isEmpty || s == '~' || s == 'null' || s == 'Null' || s == 'NULL') return null;
  if (s == 'true' || s == 'True' || s == 'TRUE') return true;
  if (s == 'false' || s == 'False' || s == 'FALSE') return false;
  if (_intRe.hasMatch(s)) {
    final v = int.tryParse(s);
    if (v != null) return v;
  }
  if (_hexIntRe.hasMatch(s)) {
    final v = int.tryParse(s.substring(2), radix: 16);
    if (v != null) return v;
  }
  if (_octIntRe.hasMatch(s)) {
    final v = int.tryParse(s.substring(2), radix: 8);
    if (v != null) return v;
  }
  if ((s.contains('.') || s.contains('e') || s.contains('E')) && _floatRe.hasMatch(s)) {
    final v = double.tryParse(s);
    if (v != null) return v;
  }
  return s;
}

/// JS `String(value)` for the scalar shapes YAML resolution produces.
String jsString(Object? value) => switch (value) {
      null => 'null',
      String s => s,
      bool b => b ? 'true' : 'false',
      int i => i.toString(),
      double d => jsNumberToString(d),
      List<Object?> list => list.map(jsString).join(','),
      Map<String, Object?> _ => '[object Object]',
      _ => value.toString(),
    };
